Allow nested at-rules for lint

maticstric opened the pull request:
In advance, sorry for the long explanation but I found a couple of issues that need a bit of explaining:

While working on Roughing It I came across a bug in the lint function: it would crash with nested at-rules because of the assumption that anything nested within an at-rule would be a selector. And just a clarification since some websites call a single @supports a nested at-rule: by nested at-rules I mean something like an @supports within an @media.

Nested at-rules are allowed according to the spec (https://www.w3.org/TR/css-conditional-3/#contents-of One example/more explanation can be found here https://www.w3.org/TR/css-conditional-3/#processing).

This code fixes that bug by recursion so technically it should work for any number of nested at-rules (tested up to three). I tested it on 200 ebooks and the `local_css_rules`, `duplicate_selectors`, and `single_selectors` variables were all kept the same with the new code so it looks like it works well. On a side note, it looks like we never use the `single_selectors` variable. If you don't think we need it, this would be a good time to get rid of it.

In terms of the code I'm no python expert so I don't know if there's any naming conventions for the recursive helper function I made. I can of course rename it if some convention exists. And feel free to change any of the comments I made but I tried my best to make them helpful.

....................................................................

And while fixing this bug I came across what I think is another bug. If you have more than one of the same top-level selector, the second will override the first value in the `local_css_rules` dictionary. Best explained by an example from https://github.com/standardebooks/maurice-leblanc_the-hollow-needle_alexander-teixeira-de-mattos/blob/master/src/epub/css/local.css:

```css
[epub|type~="z3998:postscript"]{
	margin-top: 1em;
}

[epub|type~="z3998:postscript"],
[epub|type~="z3998:recipient"] + p{
	text-indent: 0;
}
```
The `local_css_rules` dictionary after running lint here will have this as an entry `'[epub|type~="z3998:postscript"]': 'text-indent: 0;'` instead of `'[epub|type~="z3998:postscript"]': 'margin-top: 1em;text-indent: 0;'`

The only time I can really see this mattering is here:
```python
if "abbr" in selector and "nowrap" in rules:
	abbr_with_whitespace.append(selector)
```

So for a contrived worst case scenario you can imagine some css like this:
```css
#chapter-1 abbr,
#chapter-1 div{
	white-space: nowrap;
}
#chapter-1 abbr{
	background: blue;
}
```

The `local_css_rules` will contain this entry `'#chapter-1 abbr': 'background: blue;'` when it should really contain `'#chapter-1 abbr': 'white-space: nowrap;background: blue;'` and so lint wouldn't complain even though it should.

For now, I've made the new code have this same "bug" intentionally because I don't want to unintentionally break something if there's a reason for overwriting the dictionary entry. Fixing this bug would just mean replacing
```python
if top_level:
	local_css_rules[selector.selectorText] = rule.style.cssText + ";"
else:
	local_css_rules[selector.selectorText] += rule.style.cssText + ";"
```
in the new code with
```python
local_css_rules[selector.selectorText] += rule.style.cssText + ";"
```
So I can fix this very easily if you also agree that it's not the desired behavior.

acabal commented:
Which book were you working on where there's nested `@supports`? The original algorithm is pretty basic because I think needing nested `@supports` is really unusual in an ebook context... there's probably a better way to do that CSS. Do you have an example of the CSS that caused this bug?

Re. `single_selectors`, we do need it, it's tested against in your new line 445. That's how we decide if there are duplicate selectors. I don't think we need it *outside* of that function though, so it's not necessary as a return value.

Make sure to run pylint and pytest in the repo too, in case you haven't already. Instructions on how to do that are in the repo readme.

Re. the other bug, I think you're right, that's unintended. You can go ahead and fix it. Thanks!

maticstric commented:
1. The use case I wanted it for was for checking if a device supports flex AND is big enough. What brought up the issue was on the bottom of this page https://archive.org/details/innocentsabroad02twai/page/n51/mode/1up where there's the "Committee on Applications". What I would like to do is:
```css
#chapter-1 div > p{                                                                
  text-indent: 0em;                                                                
  text-align: center;                                                              
}   

@supports(display: flex){                                                        
  @media(min-width: 30em) {                                                        
    #chapter-1 div{                                                                
      display: flex;                                                               
      margin: auto;                                                                
      max-width: 30em;                                                             
      justify-content: space-around;                                               
    }                                                                              
  }                                                                                
}
```

I can imagine there would be other cases where you would want to combine some sort of @supports and @media queries and I think this is the only way to do so. And I think the code to be able to handle it doesn't add that much complexity. It's essentially doing the same thing except that when it encounters an at-rule it calls itself recursively on the rules within that rule. But I do realize that the support for @media is already fairly bad so I don't have much hope that any ereader will even recognize these nested at-rules. But for the most part SE seems to be on the side of "if a browser supports it, ereaders will eventually catch up".

2. Yea should have been more specific. I meant outside of the function. If you'd like, I can remove it from the return values if you decide to have this merged.
3. I'll look into it.
4. I'll fix it regardless of whether the nested part of this commit is merged.

acabal commented:
OK. In your CSS above `@media` is not necessary. `max-width` already does what you want for free. The div will be 100% the width of the viewport up to 30em, after which it will stop at 30em wide.

However I do think it's good to fix this bug while we have it. So if you can remove the return value, fix the other bug you found, and then make sure to run pylint and pytest and then I can merge this. Thanks!

maticstric commented:
I think there is a difference between the two but I'll move this part of the conversation over to the google group since it probably belongs there.

Other than that I have removed the return value, fixed the other bug, and ran pylint and pytest. Should be good to go. Thanks!

acabal commented:
Great work, thanks!
